sdk/core/azure_core_macros/src/tracing_new.rs

```diff
@@ -0,0 +1,312 @@
+/// 1) `Self`
+/// 1) `Arc<Self>`
+/// 1) `Result<Self, E>`
+/// 1) `Result<Arc<Self>, E>`
+///
+pub fn parse_new(attr: TokenStream, item: TokenStream) -> Result<TokenStream> {
+    if !is_new_declaration(&item) {
+        return Err(syn::Error::new(
+            item.span(),
+            INVALID_SERVICE_CLIENT_NEW_MESSAGE,
+        ));
+    }
+    let namespace_attrs: NamespaceAttribute = syn::parse2(attr)?;
+
+    let client_fn: ItemFn = syn::parse2(item.clone())?;
+
+    let vis = &client_fn.vis;
+    let ident = &client_fn.sig.ident;
+    let inputs = client_fn.sig.inputs.iter();
+    let body = client_fn.block.stmts.iter().map(|stmt| {
+        // Ensure that the body of the new function initializes the `tracer` field.
+
+        if let syn::Stmt::Expr(expr, _) = stmt {
+            if let syn::Expr::Call(c) = expr {
+                // If the expression is a call, we need to check if it is a struct initialization.
+                if c.args.len() != 1 {
+                    println!("Call expression does not have exactly one argument, emitting expression: {stmt:?}");
+                    // If the call does not have exactly one argument, just return it as is.
+                    quote! {#stmt}
+                } else if let syn::Expr::Struct(struct_body) = &c.args[0] {
+                    parse_struct_expr(namespace_attrs.client_namespace.as_str(), struct_body, stmt.to_token_stream(), true)
+                } else {
+                    println!("Call expression is not a struct, emitting expression: {stmt:?}");
+                    // If the expression is not a struct, just return it as is.
+                    stmt.to_token_stream()
+                }
+            } else if let syn::Expr::Struct(struct_body) = expr {
+                parse_struct_expr(namespace_attrs.client_namespace.as_str(), struct_body, stmt.to_token_stream(), false)
+            } else {
+                // If the expression is not a struct, just return it as is.
+                stmt.to_token_stream()
+            }
+        } else {
+            stmt.to_token_stream()
+        }
+    });
+    let output = &client_fn.sig.output;
+    Ok(quote! {
+        #vis
+        fn #ident(#(#inputs),*) #output {
+            #(#body)*
+        }
+    })
+}
+
+/// Returns true if the item at the head of the token stream is a valid service client declaration.
+fn is_new_declaration(item: &TokenStream) -> bool {
+    let item_fn: ItemFn = match syn::parse2(item.clone()) {
+        Ok(fn_item) => fn_item,
+        Err(_) => return false,
+    };
+
+    // Service clients new functions must be public.
+    if !matches!(item_fn.vis, syn::Visibility::Public(_)) {
+        return false;
+    }
+
+    // Service clients new functions must have a name that starts with `new_` or "with_".
+    if !item_fn.sig.ident.to_string().starts_with("new")
+        && !item_fn.sig.ident.to_string().starts_with("with")
```

Comment:
I'm not entirely convinced we should be enforcing this here. Rust has a fairly dominant pattern for constructor names, but I wouldn't say it's totally universal. I've seen constructors call `from_...` as well, for example. The code we generate doesn't depend on the name, so I don't think we should enforce naming here. If we want to enforce constructor naming, we can do that in linting rules, etc.

Reply:
Upon reflection, I agree.